### DISCUSSION TITLE
[discussion] Should `NO_COLOR` fully cancel out the effect of `FORCE_COLOR`?

### DISCUSSION THREAD

**Discussion by webknjaz** [donatj/force-color.org#99] [Q&A]
It looks like some projects (namely, Rich) start printing out things like bold/italic with `FORCE_COLOR` set (additionally to producing colors), but don't stop doing that when `FORCE_COLOR=1` is also present. So it results in programs producing non-colored output with ANSI sequences in stdout. When that is piped into other programs, the supposedly machine-readable output is suddenly broken and stuff explodes.

So I'd like to ask if `NO_COLOR` should fully suppress the effect of `FORCE_COLOR` or not?

_Cross-posted @ ~https://github.com/jcs/no_color/issues/267~ (that issue tracker got disabled)_

**Comment by donatj** [comment]
`FORCE_COLOR` / `NO_COLOR` are intended to affect _colored output_ only and not other ANSII control codes. Other control codes should continue to work as designed.

The reason `NO_COLOR` came to be in the first place is because some people dislike colored output and want to be able to disable it at the shell profile level. They wouldn't want or expect for it to stop other things like animated progress bars and other fancy output from working correctly.

**Comment by webknjaz** [comment]
Aha! So treating `FORCE_COLOR` as "also enable font styling" isn't correct, right?

**Comment by webknjaz** [comment]
Is this something to document in the FAQ?

**Comment by donatj** [comment]
Hmm, you've got me pondering.

It's certainly a _strange_ behaviour. I'm not sure it's my place to tell them not to though. 

I think that might kind of be an area of undefined behaviour.

**Comment by webknjaz** [comment]
I would suggest calling it out as such, in the FAQ, then…

**Comment by donatj** [comment]
I'd be curious if the https://github.com/jcs/no_color/issues/267 people have anything to say on the topic before I definitively take any sort of stance.

**Comment by webknjaz** [comment]
Absolutely! It's a good idea to have coordinated semantics of both vars, since they're effectively solving different sides of the same use case…

cc @jcs

**Comment by webknjaz** [comment]
Do we need a no-ansi.org / `NO_ANSI` now? :laughing:

**Comment by webknjaz** [comment]
> I'd be curious if the https://github.com/jcs/no_color/issues/267 people have anything to say on the topic before I definitively take any sort of stance.

UPD: it looks like @jcs disabled issues in that repository and so the issue URL returns 404 now. I haven't gotten any updates in the email before that happened, though.

@donatj sounds like it's time to make a stance, after all?

**Comment by worstpractice** [comment]
@donatj With sincere apologies for this unsolicited wall of text, I see (atleast) three potential approaches to tackling this problem. 

# 1. Paradox in, paradox out
Paradoxical configurations will produce paradoxical outcomes.

I'm being told [this is a feature and not a bug](https://en.wikipedia.org/wiki/Principle_of_explosion).

It is the sole responsibility of the end user not to express paradoxical preferences (e.g. simultaneously requesting `NO_COLOR` and `FORCE_COLOR`).

This is especially true if, as is reasonable to expect, the preferences of said end user include a strong preference against paradoxical outcomes in general.

With access to a time machine we could perhaps have "standardized" around a variable more akin to `COLOR`, which could then accept an enum of values, or indeed be absent altogether. But alas.

# 2. Interpreting `FORCE_COLOR` as the stronger of the two
Comparing the "specifications" (explainer texts) of `NO_COLOR` and `FORCE_COLOR`, I (very subjectively) get the clear impression `FORCE_COLOR` ought to win.

As the word 'force' might suggest, it comes across to me as the more imperative of the two.

While the common goal of both "standards" is the transfer of power from software _authors_ to software _users_, to the best of my understanding, `NO_COLOR` is overtly intended to be overridable, while `FORCE_COLOR` is not.

These design choices are both valid approaches towards the same goal.

However, as far as their interaction goes, `FORCE_COLOR` reads (to me) as the more nuclear option.

# 3. Boolean scrying
The best thing to do in a situation like this is to throw code at it from out of left field.

Consider [the following example code](https://www.typescriptlang.org/play/?target=99&moduleResolution=99&module=1&pretty=true&noErrorTruncation=true&declaration=false&noEmitHelpers=true&esModuleInterop=false&verbatimModuleSyntax=true&isolatedDeclarations=false&useUnknownInCatchVariables=true&noUnusedLocals=true&noUnusedParameters=true&exactOptionalPropertyTypes=true&noFallthroughCasesInSwitch=true&noUncheckedIndexedAccess=true&noImplicitOverride=true&noPropertyAccessFromIndexSignature=true&allowImportingTsExtensions=true&allowArbitraryExtensions=true&allowSyntheticDefaultImports=false&ts=5.5.0-dev.20240504&importHelpers=true&jsx=0#code/MYewdgzgLgBAYgeQEoGECiB9FCAyyYC8MA5IqptnksTAIYQyiRQDcAUE9DAHIJa74ixXvyo16jcNHZsA9ACp5MAOoALWlACmAN00AnGAHdaATwbhNMALYBLAOarYAB323YUVZc1gAJjACuEPowUCAwmgAeTnqaEAwelj6xNjEwIABmaU5QNmB2MLkhYQnhYNop4FbesNq0eja0AEYANpaFAI7+sTngAHQw8rIcUrAIAAoAKhgAktyEMADabDAwskMrxMQANKuyMIC8G4ASOzATnjAARBAuwDbpNsAaNuDnaZkABiKUyG-W1U9gsRCZyqqUiNmg3mAlgyQMstXqTVadAYtCMpleWRyeQKYF6yxIAEZtvjiFA9F1iRsTLFKSQTLTiN5EZoGeBiQBdZGSZgyBRKNQaHTBYxmNIA6z2RwwFx6Nyw0p+QLBULhKIxOLypIQFLQzIgbK5fIgfywGElbzlPSVaoweENFptMAwTrdf79QbDZgwcZTBAAVQm8yWKzWJIAtEStiSAAwM9K0ZpBBlgEDJhk+cHM1npdIcrmcVhsT1cRogECtWhgADKwD0JkN8wAFBaKmAqmAoAAuGAAbxgMVoPnAzRMiwA1ppR4VoPU8uyAPzdmcNgA+AV8mjuAL8AF8AJTdsD+KyNYIEAB8vfxfPxAxgbwgD1aEAQ6QAUoEclCYAAeGDRn4qkrBhImAZp-AzbFQGaEA9DxFY7wfJ9YlfD9oHuSwCCw-9AM0YD5SCKAsTsBgHjAKFmjSE14JWJQkITFD30-DCYEvADfnw3IwIghtoNgmiBnWGBWlgR8GJfJj0KhQ9j1PAwiGjGQVgLEIExHGtYM0OBYKbFsrTbaoADU6ntVpuFoKpuygEwXBhT4BCQGA12s2zMjIdBRGQHYMxiYAejAV8UHUPJNG7CMnJgAkD1tEAbD8C8rwQ4TNFgYBgrsUKYCPE8z3-JSENuGBmzKVt2ygYyEQdcyqhxUpLWtDs90SpLGHSywAGoiB8zQ-P+QK2vypK1gOQBxHZUdQtF0AwU1lBMbAAL0ecAMRKQjiLFSxbAcWBjA7IqdJTHbKygPcBIQlS1obIhivqgyOwWPSGvKkzmWqzROXnecSGIU6yRsKxG1+kAcBAQx9BQehNEBwaCsyRsfQwf0Jl6LjwK1RtLuxCQXM0GEEaRhZsrk9k9yantbxa4b9jGhApsMeotHvNLKwyn4VRKU8ATuU1MhKUArBcHJ-LOpLmZC+Yw263rwH6lnNHYFqYB3cJEzaOGEdmFHyLR2IMZS9bsZs3HMg17hCdk-QSbJinRbamBOpgKX-NlkKFZancKY9imqbGqtQlSEp1X8ZooAYRpRysE1Hmxejn1Q5ioTefpG32QAJPb3CmxLjySvw6ogxYyt2d3yqA1JMDSYm0vRG3sqgdgjPc3dL5p1NASvYMbdyKAcnYovymIoH8PQnSzxi0Nz9hi6LPkYD9II-HSHStHQrGW+lVwo-8iB3SGFTsAAWQAIVmNAABEMCrNAJgmWYAHEqyDa8hN6F-TajFYX96fGAyjTkJALGQi8DCNhUkAqEKByywWMuBXUMAD7H24GfC+V8b7cHvtbFYQCioqRTBAmCehoFdAxPAk+59L7XzvlWDBCFhpoFoGlaU9B4iqCtP4Bw8oYL6hCN0cw4p-BgBsC6despN7-BWp4IIMB0jNFoCRLkpcJxOlCBlBIcEKYqUendWARByaKwWF3TySB2TdjAZoPBUCExdHfi1BYtdkDGKyiAcxBDLGaGsSsKe6iRgwFHhJceLEiClnLHhastZ6x5BuqVaojcixDT2NEfWNh9BhnsDNTQXivRTQzH5eYvj45SUsH+aMFMvrEBTJIfBxAKZLmQn4hOlg2I2xgKUviegqmK27GUsIWT7hQGIDDFSQcQ6P0VqGRWsdAQwgAFb1LRAwAAJD2IJFZQl1kNJE-SZU9zK0bD0vy3ZFl7KgPuIwNgPBvHcQhBAjQpk9SgL0ao9RdaaK2U0j+VhaBOEbMVMkJgmoJV0YrBCA8h5OkeSYXoUzYpgEbMQAgP03aK33G8mAkLoWwroL4Ege5LnsjRbkDFCLYnnSkME3oME7CNiGSdIubAPZAA) (which you can run right there on the TypeScript playground to output ~~an inscrutable mess~~ a line-by-line truth table exemplifying this approach 👍).

**Comment by webknjaz** [comment]
I'd interpret `NO_COLOR` as stronger. `FORCE_COLOR` is often set globally in CIs but some programs need to be told not to output ANSI (so that the output is machine-parseable) and for that, people tend to use `NO_COLOR` on specific app invocations. Imagine the platform wants to force color where it can but the user wants to disable it in their shell — wouldn't it be desirable to listen to the user?

**Comment by worstpractice** [comment]
I'd certainly agree that erring on the side of omitting color is (AFAICT) certainly the right way to go.

**Comment by nicholaswmin** [comment]
This should be closed as completed with a final decision.

FWIW I agree with the `NO_COLOR` taking precendence.

**Comment by webknjaz** [comment]
@donatj do you think this gathered enough opinions to make the final ruling? I think that at this point, even documenting the paradox in the FAQ is better than not declaring any definitive decision.

P.S. I just want to reiterate that this is inspired by a surprising behavior of Rich where `FORCE_COLOR` adds ANSI sequences for bold and `NO_COLOR` prevents ANSI color escape sequences but doesn't cancel the rest, resulting in colorless-but-bold output that is not machine-readable at all.

**Comment by willmcgugan** [comment]
For the record, I asked the no-color dev(s) if NO_COLOR=1 disables bold and italic. They were clear it doesn't. NO_COLOR applies only to explict color ansi-codes, and *not* bold, italic or any other escape sequences.

Since they disabled the issue tracker and this isn't mentioned on no-color.org, this detail is essentially buried.

Anyway, I'm hoping my [TTY_COMPATIBLE](https://github.com/Textualize/rich/issues/2924#issuecomment-2757673602) suggestions takes off. I'm also hoping I don't have to go to the trouble of creating a website for it, which seems how terminal "standards" are created. 🤷‍♂

**Comment by webknjaz** [comment]
@donatj would it be reasonable to mention `TTY_COMPATIBLE` and the paradox in the FAQ?

**Comment by jakewilliami** [comment]
For what it's worth, I agree that `NO_COLOR` is stronger.  (This is also how UV handled it; https://github.com/astral-sh/uv/pull/3979.)  In an ideal world, it's like boolean "off" and "on", so the one **most recently set** should apply.  But most environments likely don't reliably store information around the order of variables being set.

Also, just want to say, give some love to British spelling and support `NO_COLOUR`/`FORCE_COLOUR` :smile:

**Comment by webknjaz** [comment]
@jakewilliami `os.environ` is effectively a `dict`. The declaration order wouldn't apply there.

**Comment by piotr-dobrogost** [comment]
> Anyway, I'm hoping my [TTY_COMPATIBLE](https://github.com/Textualize/rich/issues/2924#issuecomment-2757673602) suggestions takes off. I'm also hoping I don't have to go to the trouble of creating a website for it, which seems how terminal "standards" are created. 🤷‍♂

`NO_ANSI` seems like so much better name for this because it's basically self-explanatory as opposed to `TTY_COMPATIBLE` which could mean so many things…

**Comment by webknjaz** [comment]
[![](https://imgs.xkcd.com/comics/standards.png)](https://xkcd.com/927/)

**Comment by webknjaz** [comment]
@donatj I just realized I haven't seen a project in the Python ecosystem that does not have `NO_COLOR` taking preference over `FORCE_COLOR`. It might be a good idea to mention in the table on the website which of the projects have one `FORCE_COLOR` stronger and which ones have it weaker.

**Comment by donatj** [comment]
If anyone here is still interested in hashing this out, I have created an RFC #80 defining my proposed solution and am curious to hear what others think.
